src/Sarif/Readers/DeferredDictionary.cs

```diff
@@ -0,0 +1,514 @@
+        // They call the Count getter if requested by the user to avoid full enumeration until a user requests the Count.
+        public ICollection<string> Keys => new ReadOnlyCollectionAdapter<string>(() => new KeyEnumeratorAdapter<string, T>(this.GetEnumerator()), () => this.Count);
+        public ICollection<T> Values => new ReadOnlyCollectionAdapter<T>(() => new ValueEnumeratorAdapter<string, T>(this.GetEnumerator()), () => this.Count);
+
+        public int Count
+        {
+            get
+            {
+                EnsurePositionsBuilt();
+                return _itemPositions.Count;
+            }
+        }
+
+        public bool IsReadOnly => true;
+
+        #region Dictionary Mutators [not supported]
+        public void Add(string key, T value)
+        {
+            throw new NotSupportedException();
+        }
+
+        public void Add(KeyValuePair<string, T> item)
+        {
+            throw new NotSupportedException();
+        }
+
+        public void Clear()
+        {
+            throw new NotSupportedException();
+        }
+
+        public bool Remove(string key)
+        {
+            throw new NotSupportedException();
+        }
+
+        public bool Remove(KeyValuePair<string, T> item)
+        {
+            throw new NotSupportedException();
+        }
+        #endregion
+
+        public bool Contains(KeyValuePair<string, T> item)
+        {
+            T value;
+            if (!this.TryGetValue(item.Key, out value)) return false;
+            return EqualityComparer<T>.Default.Equals(value, item.Value);
+        }
+
+        public bool ContainsKey(string key)
+        {
+            T unused;
+            return this.TryGetValue(key, false, out unused);
+        }
+
+        public bool TryGetValue(string key, out T value)
+        {
+            return TryGetValue(key, true, out value);
+        }
+
+        private bool TryGetValue(string key, bool readValue, out T value)
+        {
+            value = default(T);
+            if (_stream == null) _stream = _streamProvider();
+
+            EnsurePositionsBuilt();
+
+            // Get the hash of the key
+            long keyPosition;
+            int keyHash = key.GetHashCode();
+
+            // Find the position for that hash
+            while (_itemPositions.TryGetValue(keyHash, out keyPosition))
+            {
+                _stream.Seek(keyPosition, SeekOrigin.Begin);
+
+                using (JsonTextReader reader = new JsonTextReader(new JsonObjectMemberStreamReader(_stream)))
+                {
+                    reader.CloseInput = false;
+
+                    // Read past the mock object start from JsonObjectMemberStreamReader [so JsonTextReader expects to see a PropertyName]
+                    reader.Read();
+
+                    // Get the string key for the item with this hash
+                    reader.Read();
+                    string foundKey = (string)reader.Value;
```

Comment:
ok, looking at the broader context, I see I missed the obvious point: this dictionary is associated with a single JSON object, the property names of which will not be duplicated.

---
In reply to: [220969257](https://github.com/Microsoft/sarif-sdk/pull/1100#discussion_r220969257) [](ancestors = 220969257)

Reply:
Yep - there shouldn't be duplicates, and if there are, we would only return the first one. This is similar to the Dictionary<string, FileData> would otherwise do.